### ISSUE TITLE
GraphQL addReference on Object Array

### ISSUE THREAD

**Issue by kiwimonk** [gridsome/gridsome#562]
## Summary
A way to have addReference traverse down into non root Types.
collection.addReference(fieldName, typeName) and the source-filesystem options only apply to a root Type.

#### My Input File:
```
field1: 'data'
field2: ['data1', 'data2']
field3:
- sub_field1: 'data'
  sub_field2: ['data1', 'data2']
```
#### This already works:
```
refs: {
  field1: 'typeName',
  field2: 'typeName'
}
```
#### This would be nice:
```
refs: {
  field3: {
    sub_field1: 'typeName'
    sub_field2: 'typeName'
}
```

### Motivation
References are already very powerful, but they are limited by only applying at the Root level.

**Comment by hjvedvik** [comment]
Yes, the `collection.addReference()` method is limited to root fields, unfortunately. You can, however, use the `store.createReference(typeName, id)` method to have a reference at any level if you are creating the nodes yourself.

We are also working on a new schema API which will make it possible to achieve what you want by defining the schema yourself. It would look like this:

```js
api.loadSource(({ addSchemaTypes }) => {
  addSchemaTypes(`
    type MyTypeName extends Node {
      field1: OtherTypeName
      field2: [OtherTypeName]
      field3: MyTypeNameField3
    }
    type MyTypeNameField3 {
      sub_field1: OtherTypeName
      sub_field2: [OtherTypeName]
    }
  `)
})
```

**Comment by freshyill** [comment]
@hjvedvik I see "Define and customize the internal GraphQL schema" is in progress on the roadmap. Is this planned for the next release?

**Comment by hjvedvik** [comment]
@freshyill Yes, it is :) Just working on a few other minor things that also will be in the next release these days. So hopefully it will be ready pretty soon.

**Comment by hacknug** [comment]
@hjvedvik any updated on this? I'd love to use it to improve the Airtable plugin so it handles relationships and other stuff. Put together [a CodeSandbox](https://codesandbox.io/s/gridsomegridsome-starter-default-17p1i) showing how using `addReference()` makes the relation field lose its value. You can also see how everything in the schema seems to look okay but still doesn't work.

I've been looking at core's codebase and trying random-ish things for a while but couldn't quite figure out how to solve this. If this doesn't require lots of other things to change, I could try again if anyone is able to provide some pointers on how to tackle the issue.

A little more context in a short convo we had on Discord:
- https://discordapp.com/channels/476372673564246017/504192730667155466/652134126287192076
- https://discordapp.com/channels/476372673564246017/504192730667155466/652135594595450900

**Comment by davidleininger** [comment]
@hjvedvik I'm having an issue where the references aren't getting made. I've added the following code to link one collection to another. I haven't had an issue doing this in the past, but at the moment, GraphQL is just returning `null` for every reference.

```js
api.loadSource(({ getCollection }) => {
    const restaurants = getCollection('Restaurants')
    restaurants.addReference('foodType', 'Types')
 })
```

**Comment by belvederef** [comment]
I was experiencing the same problem @davidleininger was, and I solved it by declaring my own schema. Example:
```javascript
 api.loadSource(({ addSchemaTypes }) => {
    addSchemaTypes(`
      type Comment implements Node {
        article: Article @reference(by: "id")
      }
    `);
  });
```
It was not working before specifying `@reference(by: "id")`.

Just to be explicit, your `Comment` type needs to have a field `article` with a value that matches, in this case, the `id` field on the type `Article`.

**Comment by davidleininger** [comment]
Doing it that way starts throwing errors in my build. 

For what it's worth, I just queried both pieces of content and inflated data. Also, not the cleanest, but it works. What's weird is I have other apps that are using the  `.addReference()` with no issues. It's just this one app.

**Comment by joyearnaud** [comment]
I'm not sur if it's the exact same probleme;

I'm doing in `gridsome.serve.js`:
```JS
    actions.addSchemaTypes(
      `
      type Experience implements Node @infer {
        keywords: [Keyword],
        skills: [Skill] @reference(by: "name"),
        skill: Skill @reference(by: "name"),
      }
      `
    );
```
